### PR TITLE
Add ListClients and DeleteClient operations

### LINKED ISSUE #78: Update with the admin feature
After parallaxsecond/parsec#308 is implemented, the following things need to be updated:

* In the [operations](https://parallaxsecond.github.io/parsec-book/parsec_client/operations/index.html) page we would add a paragraph about admins operations and in each of them (`ListClients` and `DeleteClient`), add the information that it is an admin operation.
* Add the new status code [here](https://parallaxsecond.github.io/parsec-book/parsec_client/status_codes.html) (see parallaxsecond/parsec-interface-rs#93)

### Threat model updates

* A9 for Spoofing and Tampering need to be updated with the new idea that someone malicious could set itself as the admin. More assets are then impacted because all clients keys can be destroyed. O-6 cover the right mitigation.
* A1 Spoofing threats are aggravated if the authentication stolen is the admin one: in that case not only admin keys are at risks but everyone's (only allows deletion of keys, not exposure/usage). I propose that we add a notice in the [AS1](https://parallaxsecond.github.io/parsec-book/parsec_security/parsec_threat_model/threat_model.html#authentication-token---as1) section noting down that the admin's authentication token is particularly sensitive as it impacts all.

### LINKED ISSUE #79: Add ListClients and DeleteClient operations
They would both be **core** operations, adressed to provider ID `0`. We should also indicate somehow that concurrency issues with those operations, that if they are executed in parallel with other operations that list/create/delete keys, the result might not be up to date or relevent.

## ListClients

Lists all clients currently having keys in the service. Opcode: 27 (`0x001B`)

## Parameters

No parameters are needed for this operation.

## Results

| Name   | Type                                 | Description             |
|--------|--------------------------------------|-------------------------|
| `clients` | Vector of String | List of clients |

## Specific response status codes

No specific response status codes returned.

## Description

This operation lists all clients that are currently storing keys in the Parsec service. The `clients` field contain a vector of the application names used by clients.

This operation necessitates **admin** privilege.

Only the clients using the same authentication method as this request will be listed. It has no impact currently as only one authentication method in the service is supported but might do if the service supports multiple.

## Contract

No contract yet.

## DeleteClient

Delete all keys a client has in the service. Opcode: 28 (`0x001C`)

## Parameters

| Name   | Type                                 | Description             |
|--------|--------------------------------------|-------------------------|
| `client` | String | Client to delete |

## Results

No values are returned by this operation.

## Specific response status codes

No specific response status codes returned.

## Description

This operation deletes all keys a client owns in all Parsec providers. The `client` parameter string must be one of the clients returned by the `ListClients` operation.

This operation necessitates **admin** privilege.

Only the clients using the same authentication method as this request will be deleted. It has no impact currently as only one authentication method in the service is supported but might do if the service supports multiple.

## Contract

No contract yet.

### PR DESCRIPTION
**Pull request by hug-dev**
Fix #79 
Fix #78

### CODE REVIEW

File: src/parsec_client/operations/delete_client.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+|----------|--------|------------------|
+| `client` | String | Client to delete |
+
+## Results
+
+No values are returned by this operation.
+
+## Specific response status codes
+
+- `AdminOperation`: this operation is an admin operation and can not be requested by a user
```
Comment (ionut-arm) [inline src/parsec_client/operations/delete_client.md:17]:
```suggestion
- `AdminOperation`: this operation is an admin operation and cannot be requested by a user
```

(same for `ListClients`)

**Review by ionut-arm — COMMENTED** [review]
Thanks! 🚀 

Left a few comments below - is your plan to merge this, then create a PR for the contracts, then create a new PR to add the links here? Or just create the contracts and then change this PR to add them?

File: src/parsec_client/operations/README.md [hunk 2]
```diff
@@ -54,6 +61,8 @@ Core operations are non-cryptographic operations supported by the core provider.
 - [ListOpcodes](list_opcodes.md)
 - [ListAuthenticators](list_authenticators.md)
 - [ListKeys](list_keys.md)
+- **admin** [ListClients](list_clients.md)
+- **admin** [DeleteClient](delete_client.md)
```
[the remark this answers is not in the document]
Comment (hug-dev) [reply src/parsec_client/operations/README.md:65]: Sure will change!

File: src/parsec_client/operations/delete_client.md [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+This operation necessitates **admin** privilege.
+
+Only the clients using the same authentication method as this request will be deleted. It has no
+impact currently as only one authentication method in the service is supported but might do if the
+service supports multiple.
+
+**Note:** this operation might return wrong results if clients' data is being modified while it
+executes. For example, if a new client is creating keys while this operation is being performed,
+this new client might not be deleted.
```
[the remark this answers is not in the document]
Comment (hug-dev) [reply src/parsec_client/operations/delete_client.md:33]: Oh yes true, that was a bad copy pasta, will change

**Comment by hug-dev** [comment]
> Left a few comments below - is your plan to merge this, then create a PR for the contracts, then create a new PR to add the links here? Or just create the contracts and then change this PR to add them?

Yes I was thinking of merging that first. There are a few other operation pages where we forgot to add their links to the contracts. We can do all of them in one batch.

File: src/parsec_client/operations/delete_client.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+
+This operation necessitates **admin** privilege.
+
+Only the clients using the same authentication method as this request will be deleted. It has no
+impact currently as only one authentication method in the service is supported but might do if the
+service supports multiple.
+
+**Note:** this operation might return wrong results if clients' data is being modified while it
+executes. For example, if the client named creates a new key while this operation is being
+performed, this client might not be deleted.
```
Comment (ionut-arm) [inline src/parsec_client/operations/delete_client.md:33]: `this client` or `this key`?

File: src/parsec_security/parsec_threat_model/threat_model.md [hunk 5]
```diff
@@ -289,15 +291,15 @@ Attacker with access to the log stream generated by the Parsec service and to th
...
-| D | Removing or altering the configuration file can lead to the Parsec service not starting or working in a broken state.         | O-6           | AS6      |
-| E | Parsing a malicious configuration file can lead to code execution at the Parsec privilege level.                              | M-5, U-3, O-6 | All      |
+|   | Description                                                                                                                                   | Mitigation    | Assets        |
+|---|-----------------------------------------------------------------------------------------------------------------------------------------------|---------------|---------------|
+| S | An adversary can spoof a configuration file, leading to the service operating in an unsecure state, ineffective state or not starting at all. | O-6           | AS4, AS5, AS6 |
+| T | The configuration file set up by the administrator might be modified, leading to the service being unsecure, ineffective or broken.           | O-6           | AS4, AS5, AS6 |
```
Comment (ionut-arm) [inline src/parsec_security/parsec_threat_model/threat_model.md:297]: Hmmm, why client data? I don't think the admin role allows you to actually view any client-related info (maybe just the client names?)

**Review by ionut-arm — COMMENTED** [review]
Couple more questions

File: src/parsec_client/operations/delete_client.md (hunk 4 above)
Comment (hug-dev) [reply src/parsec_client/operations/delete_client.md:33]: Yes `this key` seems clearer

File: src/parsec_security/parsec_threat_model/threat_model.md (hunk 5 above)
Comment (hug-dev) [reply src/parsec_security/parsec_threat_model/threat_model.md:297]: My thinking was that if the attacker can get themself to be an admin, they can call `DeleteClient` on all the clients which impacts client data.

Comment (ionut-arm) [reply src/parsec_security/parsec_threat_model/threat_model.md:297]: Hmm, I don't think it's the data that's not available, at most the keys are not available (which could be used to then read/sign/whatever the data)

Comment (hug-dev) [reply src/parsec_security/parsec_threat_model/threat_model.md:297]: Ah right! In my head it was the same client's data as in the description of the operation. I can replace with AS3 "Private Keys" for availability then?

Comment (ionut-arm) [reply src/parsec_security/parsec_threat_model/threat_model.md:297]: Yes, that's what I was thinking!

**Comment by hug-dev** [comment]
`Too Many Requests` errors

**Review by ionut-arm — APPROVED** [review]
👍🏻

**Pull request merged**